scripts/xyz_grid_support.py

```diff
@@ -302,127 +316,177 @@ def find_module(module_names):
-        xyz_grid.AxisOption("[ControlNet] Pre Resolution", identity, apply_field("control_net_pres"), confirm=confirm(int)),
-        xyz_grid.AxisOption("[ControlNet] Pre Threshold A", identity, apply_field("control_net_pthr_a"), confirm=confirm(float)),
-        xyz_grid.AxisOption("[ControlNet] Pre Threshold B", identity, apply_field("control_net_pthr_b"), confirm=confirm(float)),
-    ]
-
-    xyz_grid.axis_options.extend(extra_axis_options)
+@debug_info
+def confirm(func_or_str, p, xs):
+    if callable(func_or_str):  # func_or_str is converter
+        ListParser(xs, func_or_str, allow_blank=True)
+        return
+    elif isinstance(func_or_str, str):  # func_or_str is keyword
+        valid_data = find_dict(validation_data, func_or_str, stop=True)
+        converter = valid_data["type"]
+        exclude_list = valid_data["exclude"]() if valid_data["exclude"] else None
+        check_list = valid_data["check"]()
+        ListParser(xs, converter, allow_blank=True, exclude_list=exclude_list)
+        is_all_included(xs, check_list, allow_blank=True, stop=True)
+        return
+    else:
+        raise TypeError(
+            f"Argument must be callable or str, not {type(func_or_str).__name__}."
+        )
+
+
+def bool_(string):
+    string = str(string)
+    if string in ["None", ""]:
+        return None
+    elif string.lower() in ["true", "1"]:
+        return True
+    elif string.lower() in ["false", "0"]:
+        return False
+    else:
+        raise ValueError(f"Could not convert string to boolean: {string}")
+
+
+def choices_bool() -> list[str]:
+    return ["False", "True"]
+
+
+def choices_model():
+    controlnet.update_cn_models()
+    return list(controlnet.cn_models_names.values())
+
+
+def choices_resize_mode():
+    return [e.value for e in controlnet.ResizeMode]
+
+
+def choices_preprocessor():
+    return list(controlnet.Script().preprocessor)
+
+
+def make_excluded_list():
+    pattern = re.compile(r"\[(\w+)\]")
+    return [match.group(1) for s in choices_model() for match in pattern.finditer(s)]
+
+
+validation_data = [
+    {
+        "name": "model",
+        "type": str,
+        "check": choices_model,
+        "exclude": make_excluded_list,
+    },
+    {
+        "name": "resize_mode",
+        "type": str,
+        "check": choices_resize_mode,
+        "exclude": None,
+    },
+    {
+        "name": "preprocessor",
+        "type": str,
+        "check": choices_preprocessor,
+        "exclude": None,
+    },
+]
+
+
+def identity(x):
+    return x
+
+
+def enable_script_control():
+    shared.opts.data["control_net_allow_script_control"] = True
+
+
+def apply_field(field):
+    @debug_info
+    def apply_field_(p, x, xs, *, field=field, enclosure=apply_field):
+        enable_script_control()
+        setattr(p, field, x)
+
+    return apply_field_
+
+
+def add_axis_options(xyz_grid, extra_axis_options: list[dict]) -> None:

```

Comment:
I think it's best to write functions in the order they will be used, and place lists that are frequently modified towards the bottom. For example, like this:
```
find_module
add_axis_options
identity
enable_script_control
apply_field
confirm
bool_
choices_for
make_excluded_list
validation_data
extra_axis_options
```

Reply:
Added in comments for clarity for new coders. Good idea?